sql, client: auto-retry pushed txns that can't be refreshed

andreimatei opened the pull request:
... if it's still possible for the SQL connExecutor to retry them (i.e.
we haven't sent results to the client for the respective txn yet).

This patch re-introduces code we used to have, but was deleted in #21140.
The motivation for deleting the code was that retriable errors were
supposed to be rarer because of the "read refreshing" mechanism, but
that mechanism doesn't apply to transactions that "observe their commit
timestamp". A txn can observe its timestamp by calling
cluster_logical_timestamp(), or in other ways. In particular, schema
change txns always observe their ts.

Fixes #22933

Release note: Retriable errors on schema change operations are less
likely to be returned to clients; more operations are retried
internally.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24031)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/internal/client/txn.go, line 1239 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24031#-L7zoaMtC7dd-2yK3941:-L7zoaMtC7dd-2yK3942:bihtuve) ([raw file](https://github.com/cockroachdb/cockroach/blob/24042bcad9c79634a9f2a5e0e4c780efe8bee25d/pkg/internal/client/txn.go#L1239)):*
> ```Go
> 
> 	origTimestamp := txn.Proto().OrigTimestamp
> 	origTimestamp.Forward(txn.Proto().RefreshedTimestamp)
> ```

This local `origTimestamp` variable is unused. Should the next line refer to `origTimestamp` instead of `txn.Proto().OrigTimestamp`? (If so, we need a test where this matters too)

---

*[pkg/sql/logictest/logic.go, line 810 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24031#-L7zpYRH-YTT5Yz8l1AZ:-L7zpYRH-YTT5Yz8l1A_:b-lai8u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/24042bcad9c79634a9f2a5e0e4c780efe8bee25d/pkg/sql/logictest/logic.go#L810)):*
> ```Go
> 	// connection.
> 	inheritDatabase     databaseOpt = false
> 	dontInheritDatabase databaseOpt = true
> ```

It's confusing to have true/false aliases where the `dont` version is `true`. Better to flip these and explicitly init the `var dbOpt databaseOpt` as `dbOpt := inheritDatabase` (or just get rid of the type completely; I'm not sure it adds any clarity). 

The "inheritance" mechanism is clumsy. The database is (nearly) always `test`, so can we just put that in our PGURL below so that `user` always starts you out on the `test` database and tests that need something different can set it? 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24031#-:-L7zu8WJBci7uyeHdd78:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
sql, client: auto-retry pushed txns that can't be refreshed

... if it's still possible for the SQL connExecutor to retry them (i.e.
we haven't sent results to the client for the respective txn yet).

This patch re-introduces code we used to have, but was deleted in #21140.
The motivation for deleting the code was that retriable errors were
supposed to be rarer because of the "read refreshing" mechanism, but
that mechanism doesn't apply to transactions that "observe their commit
timestamp". A txn can observe its timestamp by calling
cluster_logical_timestamp(), or in other ways. In particular, schema
change txns always observe their ts.

Fixes #22933

Release note: Retriable errors on schema change operations are less
likely to be returned to clients; more operations are retried
internally.

andreimatei committed:
sql: make logictest conns start with the "test" db

Before this patch, the logictest used contrived logic to set a database
on new connections matching the current database on the connection that
it was switching away to. It was too complicated and relied on running
SQL statements, which had various side-effects.

Release note: None

andreimatei committed:
sql: cleanup after logic test

Release note: None

andreimatei commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/internal/client/txn.go, line 1239 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24031#-L7zoaMtC7dd-2yK3941:-L8-CfkDCvf7_2gZ9xKf:bhdjcf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/24042bcad9c79634a9f2a5e0e4c780efe8bee25d/pkg/internal/client/txn.go#L1239)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This local `origTimestamp` variable is unused. Should the next line refer to `origTimestamp` instead of `txn.Proto().OrigTimestamp`? (If so, we need a test where this matters too)
</blockquote></details>

good catch, changed.

As discussed, I'll leave this without a test as it's not clear whether there can be a refreshed ts before commit, at least currently.

---

*[pkg/sql/logictest/logic.go, line 810 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24031#-L7zpYRH-YTT5Yz8l1AZ:-L8-0zRs3aYqnOqFPKMu:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/24042bcad9c79634a9f2a5e0e4c780efe8bee25d/pkg/sql/logictest/logic.go#L810)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's confusing to have true/false aliases where the `dont` version is `true`. Better to flip these and explicitly init the `var dbOpt databaseOpt` as `dbOpt := inheritDatabase` (or just get rid of the type completely; I'm not sure it adds any clarity). 

The "inheritance" mechanism is clumsy. The database is (nearly) always `test`, so can we just put that in our PGURL below so that `user` always starts you out on the `test` database and tests that need something different can set it? 
</blockquote></details>

done

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24031)*
<!-- Sent from Reviewable.io -->
